### PR TITLE
Detect database file changes using hash

### LINKED ISSUE #2230: "closing without saving" prompt when database was not modified and was on external storage unmounted without closing keepass
Hi,
My kbdx files resides on an external USB key.

## Current Behavior

Keepassxc in set to:
- automatically save after every change
- don't mark database as modified for non-data changes

"automatically reload the database when modified externally" is unchecked so the database is not shared at the same moment with other programs

In that case if I forgot to close keepassxc and dismount the external media, a prompt "files has changed, do you want to load the changes?" appears, Choosing "no" and closing without saving works (please remember that is set to "automatically save after every change")..

## Expected Behavior
In my opinion this leads to confusion as until there is no change made to the database keepassxc should just close right away

Sincerely
Luca

## Debug Info
KeePassXC - Version 2.3.4
Revision: 6fe821c

Libraries:
- Qt 5.7.1
- libgcrypt 1.7.6-beta

Operating system: Debian GNU/Linux 9 (stretch)
CPU architecture: x86_64
Kernel: linux 4.9.0-8-amd64

Enabled extensions:
- Auto-Type
- YubiKey

**these also appears but they are not enabled:**
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent

### LINKED ISSUE #2389: Detect changes to database file prior to save and prompt to merge
## Expected Behavior

If the underlying database file changes for whatever reason, KeePassXC should detect this and prompt to merge those changes prior to saving.

## Current Behavior

This works sometimes using the file watcher construct. However, it often fails for USB drives and network shares.

## Possible Solution

Store a hash of the database file when opening and after every save. Compare this hash to one developed just prior to saving and trigger actions if they differ.

## Steps to Reproduce (for bugs)

1. Create database, store on SMB share
2. Open database on another computer, make changes
3. Make changes on original computer, save
4. Data is lost

### LINKED ISSUE #2536: External Changes on Samba Share not recognized on Linux
We work in a shared network environment with four users using one OSX, two Windows 10 and one Ubuntu 18.10 (me). Keepass databases are stored on a NAS that is accessible via Samba shares. We verified, that the KeepassXC instances on the Windows 10 machines properly detect external changes of the KeePass database (changes were made on Windows and Linux) while my instance on Ubuntu Linux does not. I played with the settings, disabled safe saving and auto-saving, but my KeepassXC never updated and always overwrites external changes without prompting or merging.

I tested this with current stable release 2.3.4 and with a custom compile from the develop branch. Both builds behave exactly the same. Fix from #1799 does not fix this issue on Samba shares.

## Expected Behavior
Like on Windows 10 systems, I expect KeepassXC to detect external changes of the database and do a reload of the database.

## Current Behavior
External changes are not detected when the database is located on a Samba share

## Steps to Reproduce
1. Put a kdbx file on a Samba share
2. Open the kdbx File on a Linux workstation and a Windows 10 Workstation
3. Change an entry on the Windows workstation. (notice the Linux workstation does not update)
4. Change a different entry on the Linux workstation and safe the kdbx file.
5. KeepassXC on the Windows workstation will reload the file and the changes there will disappear.

## Debug Info
KeePassXC - Version 2.3.4
Revision: 6fe821c

Libraries:
- Qt 5.11.1
- libgcrypt 1.8.3

Operating system: Ubuntu 18.10
CPU architecture: x86_64
Kernel: linux 4.18.0-11-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

### LINKED ISSUE #3506: Could not open the database file while attempting to autoreload
I am getting the error "Could not open the new database file while attempting to autoreload." when someone else saves the database and it also is open on my computer.

Background information:
I am using KeepassXC 2.4.3 on an Windows 10 Build 1903 system. The database is located on a mapped network drive hosted on a Ubuntu Samba server, available for me and my colleagues. Most of my colleagues are using the original Keepass2 in the latest version on Windows too. We all have the database open all the day and at some time someone changes and entry, adds one, deletes one or whatever you can imagine.

Then KeepassXC recognizes that something has changed, tries to autoreload but fails with a red message telling me it was not able to open the database. Then after a few seconds the error message disappears. However it still did not reload the database. The changes are not available. I have to close the database and reload it again to see the changes. That is not what I was expecting to happen.

I don't want to establish KeepassXC while uninstalling Keepass2 on all computers until I have worked with it a little while. I mostly want to use it because of the better integration in Firefox and Chrome.

## Expected Behavior
If KeepassXC is telling me that it is not able to open the database while attempting to autoreload I expect it to try it a few more times instead of just hide the error after a few seconds and pretend that nothing did happen.

## Current Behavior
When someone else changes the database I get an error that the database can not be opened. After a few seconds the error disappears and I don't get the changes.

## Steps to Reproduce
1. Install Keepass2 on one Windows PC.
2. Create a database on a mapped network drive.
3. Install KeepassXC on an other computer.
4. Open the database on both of them.
5. Change an entry with Keepass2 and save.
6. KeepassXC should show the error, hide it again and do nothing more.

## Context
See above.

## Debug Info
KeePassXC - 2.4.3
Revision: 5d6ef0c

Libraries:
- libgcrypt 1.8.4

Operating system: Windows 10 (10.0)
CPU architecture: x86_64
Kernel: winnt 10.0.18362

Enabled extensions:
- Auto-Tyoe
- Browser Integration
- SSH Agent
- KeeShare (signed and unsiged sharing)
- YubiKey

### PR DESCRIPTION
**Pull request by droidmonkey**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
**Cleanup Database Save Functions**
* Make a clear distinction between saving to the existing file path and saving to a new file path
* Use proper save function calls in CLI

**Move FileWatcher into database class**
* Fix #3506
* Fix #2389
* Fix #2536
* Fix #2230

Every database that has been opened now watch's it's own file. This allows the database class to manage file changes and detect fail conditions during saving. Additionally, all stakeholders of the database can listen for the database file changed notification and respond accordingly.

Performed significant cleanup of the autoreload code within DatabaseWidget. Fixed several issues with handling changes due to merging, not merging, and other scenarios while reloading.

Prevent database saves to the same file if there are changes on disk that have not been merged with the open database.

**Improve database and CLI tests**

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested manually and improved test cases.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/core/Database.cpp [hunk 1]
```diff
@@ -150,69 +153,97 @@ bool Database::open(const QString& filePath, QSharedPointer<const CompositeKey>
...
 }
 
 /**
- * Save the database to a file.
+ * Save the database to a specific file.
  *
- * This function uses QTemporaryFile instead of QSaveFile due to a bug
- * in Qt (https://bugreports.qt.io/browse/QTBUG-57299) that may prevent
- * the QSaveFile from renaming itself when using Dropbox, Drive, or OneDrive.
+ * If atmoic is false, this function uses QTemporaryFile instead of QSaveFile
```
Comment (louib) [inline src/core/Database.cpp]: @droidmonkey atomic*

File: src/core/Database.cpp [hunk 2]
```diff
@@ -150,69 +153,97 @@ bool Database::open(const QString& filePath, QSharedPointer<const CompositeKey>
...
+            if (error) {
+                *error = tr("Database file has unmerged changes.");
+            }
+            return false;
         }
-        return false;
     }
 
+    // Clear read-only flag
+    setReadOnly(false);
```
Comment (louib) [inline src/core/Database.cpp]: @droidmonkey why are we clearing the read-only flag here? Isn't there a chance that we save a read-only db but to a different file?

File: tests/TestCli.cpp [hunk 3]
```diff
@@ -269,17 +265,25 @@ void TestCli::testAdd()
...
     Utils::Test::setNextPassword("a");
     Utils::Test::setNextPassword("newpassword");
     addCmd.execute(
         {"add", "-u", "newuser2", "--url", "https://example.net/", "-p", m_dbFile->fileName(), "/newuser-entry2"});
+    m_stdoutFile->seek(pos);
+    m_stderrFile->seek(posErr);
+    m_stdoutFile->readLine(); // skip password prompt
+    m_stdoutFile->readLine(); // skip password input
+    QCOMPARE(m_stderrFile->readAll(), QByteArray(""));
+    QCOMPARE(m_stdoutFile->readAll(), QByteArray("Successfully added entry newuser-entry2.\n"));
```
Comment (louib) [inline tests/TestCli.cpp]: @droidmonkey :+1:

File: tests/TestCli.cpp [hunk 4]
```diff
@@ -575,7 +579,7 @@ void TestCli::testCreate()
     QVERIFY(db2);
 
     // Testing with existing keyfile
-    QString databaseFilename3 = testDir->path() + "testCreate3.kdbx";
+    QString databaseFilename3 = testDir->path() + "/testCreate3.kdbx";
```
Comment (louib) [inline tests/TestCli.cpp]: @droidmonkey :interrobang: how was it working without the extra slash?

File: src/core/Database.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/core/Database.cpp]: Once you save it to a new file it is by definition no longer read only

File: tests/TestCli.cpp (hunk 4 above)
Comment (droidmonkey) [reply tests/TestCli.cpp]: It was creating a file in the temp directory, not in the temporary directory made from this class.

**Comment by droidmonkey** [comment]
Changes made and fixed tests

**Comment by droidmonkey** [comment]
@phoerious this is ready for your review

File: src/core/FileWatcher.h [hunk 5]
```diff
@@ -23,34 +23,39 @@
 #include <QTimer>
 #include <QVariant>
 
-class DelayingFileWatcher : public QObject
+class FileWatcher : public QObject
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/core/FileWatcher.h]: The FileWatcher denounces all who oppose it

File: src/core/Database.cpp [hunk 6]
```diff
@@ -150,69 +153,97 @@ bool Database::open(const QString& filePath, QSharedPointer<const CompositeKey>
...
+    if (ok) {
+        setFilePath(filePath);
+    } else {
+        // Saving failed
+        markAsModified();
+    }
 
+    m_fileWatcher->start(canonicalFilePath);
+    return ok;
+}
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/core/Database.cpp]: good point

**Comment by droidmonkey** [comment]
Changes made, I recommend a merge commit for this one since a lot of fundamental code was changed.

**Comment by droidmonkey** [comment]
I will verify but pretty sure they do.

**Comment by FlorentDotMe** [comment]
That's a great improvement ! Thank you for your work. As the new release isn't published yet, I don't really want to open a new issue. Just to point out an unexpected behavior with this feature.

### Expected Behavior
User A and User B are simultaneously using the same keepass file.

User A is editing the new entry E1. At the same time, User B is editing a new entry E2. User A is the first to validate his entry E1 while User B is still editing E2. Database is updated. Then User B validates his entry E2. Database is updated. Both users see E1 and E2 in the database.

### Current Behavior
User A is editing the new entry E1. At the same time, User B is adding a new entry E2. User A is the first to validate his entry E1 while User B is still editing E2. Database is updated. Then User B validates his entry E2. **Database is NOT updated.** Finally, only E1 was added to the database and E2 is lost.

### Debug Info
KeePassXC - Version 2.5.0-snapshot
Build Type: Snapshot
Revision: 4cc06f9

Qt 5.11.3
Debugging mode is disabled.

Operating system: Debian GNU/Linux 10 (buster)
CPU architecture: x86_64
Kernel: linux 4.19.0-5-amd64

Enabled extensions:
- Auto-Type

Cryptographic libraries:
 libgcrypt 1.8.4

**Comment by droidmonkey** [comment]
This scenario would occur without this change. Please open a new issue for this and a more refined sequence of events please.

**Pull request merged**